### PR TITLE
[libc++][RFC] Always define internal feature test macros

### PR DESCRIPTION
**Pull request by philnik777**
Currently, the library-internal feature test macros are only defined if the feature is not available, and always have the prefix `_LIBCPP_HAS_NO_`. This patch changes that, so that they are always defined and have the prefix `_LIBCPP_HAS_` instead. This changes the canonical use of these macros to `#if _LIBCPP_HAS_FEATURE`, which means that using an undefined macro (e.g. due to a missing include) is diagnosed now. While this is rather unlikely currently, a similar change in `<__configuration/availability.h>` caught a few bugs. This also improves readability, since it removes the double-negation of `#ifndef _LIBCPP_HAS_NO_FEATURE`.

The current patch only touches the macros defined in `<__config>`. If people are happy with this approach, I'll make a follow-up PR to also change the macros defined in `<__config_site>`.

### CODE REVIEW

**Review by ldionne — COMMENTED** [review]
I generally like this change.

On the one hand, I love getting rid of the double negative, the fact that we'll trigger `-Wundef` and the fact that we make it a bit clearer what is intended to be set by the library and not by users.

On the other hand, this change makes it look a bit like `_LIBCPP_HAS_EXCEPTION` is an "optional" feature just like having ASAN turned on. I guess they are actually the same and I just don't think about them that way, but it could be that `_LIBCPP_HAS_NO_EXCEPTIONS` has been there for so long and is so prevalent that it has a special place in my mind.

I am generally happy with this patch, but I expect it will require a bit of adjustment for downstreams. Pinging @llvm/libcxx-vendors for awareness.

I'm eager to know what other folks think about this, so I won't approve just yet. Let's let this bake for at least a few days.

**Comment by rupprecht** [comment]
What kinds of adjustments downstream do you expect this will require?

Maybe it's just something historical, but I see some `-D_LIBCPP_HAS_NO_<x>` usage. Does this PR mean we should use `-D_LIBCPP_HAS_NO_FOO -D_LIBCPP_HAS_FOO=0` as a way to be compatible before & after this change?

OTOH, it seems suspicious that we're setting these in the first place. If the conditions in `__config` lead to libc++ defining `#define _LIBCPP_HAS_FOO 1` but we have `-D_LIBCPP_HAS_FOO=0` as a build flag, that would break due to macro redefinition w/ a different value, right? Are there any conditions here where it would make sense for `__config` to only define it if not already defined?

One common option I see being set is `_LIBCPP_HAS_NO_THREADS`, but that isn't being changed here. Did you miss that one, or is intentional that only a subset of the `HAS_NO` vars are being changed?

+1 to this patch being an improvement, especially with the removal of double-negative as mentioned.

**Comment by ldionne** [comment]
> What kinds of adjustments downstream do you expect this will require?

Since this touches a lot of widely used macros, any downstream-only code that is conditionalized on exceptions would likely use `_LIBCPP_HAS_NO_EXCEPTIONS` and would need to change. That's just an example, but you get the idea.

> 
> Maybe it's just something historical, but I see some `-D_LIBCPP_HAS_NO_<x>` usage. Does this PR mean we should use `-D_LIBCPP_HAS_NO_FOO -D_LIBCPP_HAS_FOO=0` as a way to be compatible before & after this change?

If you can share which `FOO`s you found these usages for, that could also be enlightening.

> 
> OTOH, it seems suspicious that we're setting these in the first place. If the conditions in `__config` lead to libc++ defining `#define _LIBCPP_HAS_FOO 1` but we have `-D_LIBCPP_HAS_FOO=0` as a build flag, that would break due to macro redefinition w/ a different value, right? Are there any conditions here where it would make sense for `__config` to only define it if not already defined?

Exactly, right. So libc++ has a notion of "user configurable settings", but not all the `_LIBCPP_HAS_NO_FOO` fall into that category. Unfortunately, whether something is intended to be user-configurable or not is usually not well documented and it's mostly something that the developers know.

In the general case, users are not expected to define any `_LIBCPP` macro unless it's documented as a public configurable thing. In practice, people probably often define macros in ways we haven't thought about.

> 
> One common option I see being set is `_LIBCPP_HAS_NO_THREADS`, but that isn't being changed here. Did you miss that one, or is intentional that only a subset of the `HAS_NO` vars are being changed?

It looks like an oversight on @philnik777 's part. It also seems incorrect for users to be setting that, since this should be done via `__config_site` at configuration-time of the library.

IMO this patch can be a first step towards formalizing what is and isn't a user-configurable setting. We could potentially use a different naming convention to make it more obvious what is configurable. Or we could potentially do something like

```
#ifdef _LIBCPP_HAS_EXCEPTIONS
#  error "_LIBCPP_HAS_EXCEPTIONS is not supposed to be user-configured, it's supposed to be inferred by libc++"
#endif

#if __has_feature(__cxx_exceptions)
#  define _LIBCPP_HAS_EXCEPTIONS 1
#else
#  define _LIBCPP_HAS_EXCEPTIONS 0
#endif
```

That would make it clear what we intend for users to set and what we don't. But any such thing should be left for a different patch.

**Comment by rupprecht** [comment]
> > What kinds of adjustments downstream do you expect this will require?
> 
> Since this touches a lot of widely used macros, any downstream-only code that is conditionalized on exceptions would likely use `_LIBCPP_HAS_NO_EXCEPTIONS` and would need to change. That's just an example, but you get the idea.

Oh, I was only looking at build configurations, I completely forgot about that. I see quite a bit more when I expand the scope for things like that, e.g. https://github.com/abseil/abseil-cpp/blob/192e959b16809f751d565b53a949b21129d904fb/absl/hash/internal/hash.h#L73

Probably still tractable, but not as trivial as I thought.

It may help to temporarily leave in the old definitions for downstream use, even if libc++ itself does not use it, i.e.

```
#if __has_feature(__cxx_exceptions)
#  define _LIBCPP_HAS_EXCEPTIONS 1
#else
#  define _LIBCPP_HAS_EXCEPTIONS 0
#  ifdef _LIBCPP_DEFINE_DEPRECATED_FEATURE_TEST_MACROS  // <-- remove after an LLVM release or whatever
#    define _LIBCPP_HAS_NO_EXCEPTIONS
#  endif
#endif
```

(we may just need this for some popular subset of macros)

> 
> > Maybe it's just something historical, but I see some `-D_LIBCPP_HAS_NO_<x>` usage. Does this PR mean we should use `-D_LIBCPP_HAS_NO_FOO -D_LIBCPP_HAS_FOO=0` as a way to be compatible before & after this change?
> 
> If you can share which `FOO`s you found these usages for, that could also be enlightening.

_LIBCPP_HAS_NO_FILESYSTEM_LIBRARY
_LIBCPP_HAS_NO_INT128
_LIBCPP_HAS_NO_LIBRARY_ALIGNED_ALLOCATION
_LIBCPP_HAS_NO_LOCALIZATION
_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER
_LIBCPP_HAS_NO_STDIN
_LIBCPP_HAS_NO_STDOUT
_LIBCPP_HAS_NO_THREADS
_LIBCPP_HAS_NO_TIME_ZONE_DATABASE
_LIBCPP_HAS_NO_UNICODE
_LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS

> > One common option I see being set is `_LIBCPP_HAS_NO_THREADS`, but that isn't being changed here. Did you miss that one, or is intentional that only a subset of the `HAS_NO` vars are being changed?
> 
> It looks like an oversight on @philnik777 's part. It also seems incorrect for users to be setting that, since this should be done via `__config_site` at configuration-time of the library.

It looks like all the usage is in build macros to generate a `__config_site` file.

**Comment by philnik777** [comment]
As I said in the commit message, this only changes FTMs configured in `<__config>`. `<__config_site>` will be a follow-up patch. Re. some of the macros:

> _LIBCPP_HAS_NO_FILESYSTEM_LIBRARY
> _LIBCPP_HAS_NO_STDIN
> _LIBCPP_HAS_NO_STDOUT

We don't use these macro names anymore, so any code that checks the macro is already broken.

> _LIBCPP_HAS_NO_LIBRARY_ALIGNED_ALLOCATION
> _LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS
> _LIBCPP_HAS_NO_INT128

I don't know why anybody would ever want to define that. Maybe some workaround for a bug?

> _LIBCPP_HAS_NO_LOCALIZATION
> _LIBCPP_HAS_NO_THREADS
> _LIBCPP_HAS_NO_TIME_ZONE_DATABASE
> _LIBCPP_HAS_NO_UNICODE

I'd be really interested what the reasoning for these is. Maybe not properly configured libc++ builds and people try to work around it?

> _LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER

I don't plan to change that one, but it should really only be set by our tests. OTOH I don't really care if people can't configure their warnings correctly because they define library-internal macros.

**Comment by iains** [comment]
> > _LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS

GCC does not implement the availability attribute (yet) I've been working on it - but it won't be in GCC-14 - so this would potentially affect GCC on Apple platforms.

> > _LIBCPP_HAS_NO_INT128

No 32b targets left?

**Comment by philnik777** [comment]
> > > _LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS
> 
> GCC does not implement the availability attribute (yet) I've been working on it - but it won't be in GCC-14 - so this would potentially affect GCC on Apple platforms.
> 
> > > _LIBCPP_HAS_NO_INT128
> 
> No 32b targets left?

Availability annotations are disabled automatically on non-Clang compilers and __int128 availability is also detected through `__SIZEOF_INT128__`. There should be no need to set them explicitly.

**Comment by ldionne** [comment]
> 
> Oh, I was only looking at build configurations, I completely forgot about that. I see quite a bit more when I expand the scope for things like that, e.g. https://github.com/abseil/abseil-cpp/blob/192e959b16809f751d565b53a949b21129d904fb/absl/hash/internal/hash.h#L73

This specific usage looks wrong. I think it was a workaround for the fact that feature-test macros didn't take into account availability on the deployment target, but this has been fixed now. Checking for the feature-test macro only should be sufficient.

**Comment by EricWF** [comment]
I have a preference for the existing mechanisms. They've not caused my any trouble in my time on the project.

I think the existing mechanisms encourage/enforce the use of negative feature test macro names (ex. `_LIBCPP_HAS_NO_EXCEPTIONS`).  I believe this naming scheme helps avoid misspelling/misconfiguration bugs,
since a misspelled guard will almost certainly cause a compilation failure. I worry the proposed approach isn't as robust against this.

This would change very long standing existing practice, which according to my recollection hasn't caused any major problems. Generally the macros are specified in the negative form (ex. _LIBCPP_HAS_NO_EXCEPTIONS), so they're only defined when a feature isn't available.

Because the current practice is so long standing, I would like this change to come with a clang-tidy that looked for misapplication of `#ifdef` to libc++ macros.

**Comment by philnik777** [comment]
> I have a preference for the existing mechanisms. They've not caused my any trouble in my time on the project.
> 
> I think the existing mechanisms encourage/enforce the use of negative feature test macro names (ex. `_LIBCPP_HAS_NO_EXCEPTIONS`). I believe this naming scheme helps avoid misspelling/misconfiguration bugs, since a misspelled guard will almost certainly cause a compilation failure. I worry the proposed approach isn't as robust against this.

I'm not sure how the current mechanism is more robust. A misspelling might not be caught other than by testing the actual code (eg. `#ifdef _LIBCPP_HAS_NO_EXCEPTION` misspelled). The new mechanism will catch all misspellings at compile time (possibly even in your editor), since we use `-Werror -Wundef`. As I said in the description, a very similar change in `<__availability>` caught real bugs.

> This would change very long standing existing practice, which according to my recollection hasn't caused any major problems. Generally the macros are specified in the negative form (ex. _LIBCPP_HAS_NO_EXCEPTIONS), so they're only defined when a feature isn't available.
> 
> Because the current practice is so long standing, I would like this change to come with a clang-tidy that looked for misapplication of `#ifdef` to libc++ macros.

While the practice is long-standing, the names of the individual macros changed before (cf65d27, b22aa3d, 5c40c99, 66a562d), without any major fallout I'm aware of. So I don't think a clang-tidy check pulls its weight here. A general check for internal macros might be nice, but is definitely out of scope for this change.

**Comment by EricWF** [comment]
> > I have a preference for the existing mechanisms. They've not caused my any trouble in my time on the project.
> > I think the existing mechanisms encourage/enforce the use of negative feature test macro names (ex. `_LIBCPP_HAS_NO_EXCEPTIONS`). I believe this naming scheme helps avoid misspelling/misconfiguration bugs, since a misspelled guard will almost certainly cause a compilation failure. I worry the proposed approach isn't as robust against this.
> 
> I'm not sure how the current mechanism is more robust. A misspelling might not be caught other than by testing the actual code (eg. `#ifdef _LIBCPP_HAS_NO_EXCEPTION` misspelled). 

Yes, like with `-Wundef` you need to compile the code.

That  said, lets say had

```c++
#ifndef _LIBCPP_HAS_NO_EXXXXCEPTIONS
throw 42;
#endif
```

This gets caught because the code fails to compile. If the code successfully compiles, that means there's some other bugs going on that needs fixing. So it catches both misspellings and misconfigurations.

Note that the negative naming of the macro matters here. It's And is important even with the new approach.

With this change, we open up a bunch of new failure modes.

I'm more ok with changing to 1 or 0 than I am removing the negative feature test names. Did the `<__availability>` bugs you referenced use negative feature test names? Those are important, and I would consider blocking this change if it meant using positive feature tests.

Let me know if I need to do a better job of explaining why positive names are less desirable than negative ones.

> As I said in the description, a very similar change in `<__availability>` caught real bugs.

Would you mind referencing the change? 

> 
> > This would change very long standing existing practice, which according to my recollection hasn't caused any major problems. Generally the macros are specified in the negative form (ex. _LIBCPP_HAS_NO_EXCEPTIONS), so they're only defined when a feature isn't available.
> > Because the current practice is so long standing, I would like this change to come with a clang-tidy that looked for misapplication of `#ifdef` to libc++ macros.
> 
> While the practice is long-standing, the names of the individual macros changed before ([cf65d27](https://github.com/llvm/llvm-project/commit/cf65d275ac048dfb781c7302c6c82c3500184eb1), b22aa3d, 5c40c99, 66a562d), without any major fallout I'm aware of. So I don't think a clang-tidy check pulls its weight here. A general check for internal macros might be nice, but is definitely out of scope for this change.

You're renaming the macros because macro usage is bug prone. I'm don't agree with the claim, and I don't think it's out of scope, given that this change is about catching macro bugs and is otherwise NFC. 

Let me know how you would like to proceed.

**Comment by ldionne** [comment]
> I'm more ok with changing to 1 or 0 than I am removing the negative feature test names. Did the `<__availability>` bugs you referenced use negative feature test names? Those are important, and I would consider blocking this change if it meant using positive feature tests.
> 
> Let me know if I need to do a better job of explaining why positive names are less desirable than negative ones.
> 

I would like to better understand this aspect. My understanding is that we've been using negative "configuration macros" mainly because when we use `#ifdef` as a branching mechanism, it's more natural that the state where nothing is defined is the default configuration of the library. And since most (but not all) of our "configuration macros" are removing things from the default configuration (e.g. `-fno-exceptions`, `-fno-rtti`, etc), it makes sense that these macros would be negative.

However, note that this pattern doesn't always make sense. For example, @philnik777  pointed out to me that we have `_LIBCPP_HAS_NO_ASAN`. Our default configuration of the library doesn't assume ASAN, and so IMO it looks weird that our default configuration would be what's inside `#ifndef _LIBCPP_HAS_NO_ASAN`. Instead, it would make more sense to say `#ifdef _LIBCPP_HAS_ASAN`.

This patch basically puts all the configuration macros on the "same level" and uses a consistent mechanism to handle all of them.

There's still the question of whether we want to have

```c++
#if !_LIBCPP_HAS_NO_EXCEPTIONS
...
#endif
```

versus

```c++
#if _LIBCPP_HAS_EXCEPTIONS
...
#endif
```

IMO, avoiding the double-negative in this case makes things a lot easier to understand.

To summarize: I agree that negative macros make more sense in a `#ifndef` world, however I think their benefit is lost when we move to `0-1` macros instead. I'd love to know your thoughts on that.

**Comment by zibi2** [comment]
> IMO, avoiding the double-negative in this case makes things a lot easier to understand.

I agree, I don't like double negation. It's hard to read. If the change makes our lives easier we should do it.

**Comment by EricWF** [comment]
I agree with most of what @ldionne said. There in benefit to the negative macro formulation that is lost.
That benefit is making it harder to accidentally misuse the macro in a way that hides its misuse.

This can be mitigated by testing, and by tooling. If we're going to go this path, I think we should change the naming scheme to remove `_LIBCPP_HAS`. These sorts of names have a history of being defined/undefined, but never defined to `0`. To avoid confusion, I think switching to a different naming scheme (any different naming system), would be valuable.

I might suggest `_LIBCPP_FEATURE_` ?

**Comment by EricWF** [comment]
> While the practice is long-standing, the names of the individual macros changed before ([cf65d27](https://github.com/llvm/llvm-project/commit/cf65d275ac048dfb781c7302c6c82c3500184eb1), [b22aa3d](https://github.com/llvm/llvm-project/commit/b22aa3d74f276afbfbd10953de68e3b4f8cef907), [5c40c99](https://github.com/llvm/llvm-project/commit/5c40c994c329aa7e521c53cbc893426c90203941), [66a562d](https://github.com/llvm/llvm-project/commit/66a562d22e708ba40b8443b58e504ac3f983ba59)), without any major fallout I'm aware of. So I don't think a clang-tidy check pulls its weight here. A general check for internal macros might be nice, but is definitely out of scope for this change.

I don't believe any of those changes changed the way the macro works. They simply re-named things. So I wouldn't expect them to have "fallout" in the way I'm suggesting.

Also, if the purpose of this change is entirely for bug prevention & macro misuse, then a tool to assist with that is well in scope.

**Comment by ldionne** [comment]
In light of the discussion we had in the libc++ monthly, I think we could update this patch. @philnik777

File: libcxx/include/new [hunk 1]
```diff
@@ -103,16 +103,22 @@ void  operator delete[](void* ptr, void*) noexcept;
...
-#  define _LIBCPP_HAS_NO_LANGUAGE_SIZED_DEALLOCATION
+#if defined(__cpp_sized_deallocation) && __cpp_sized_deallocation >= 201309L
+#  define _LIBCPP_HAS_LANGUAGE_SIZED_DEALLOCATION 1
+#else
+#  define _LIBCPP_HAS_LANGUAGE_SIZED_DEALLOCATION 0
 #endif
 
-#if !defined(_LIBCPP_BUILDING_LIBRARY) && _LIBCPP_STD_VER < 14 && defined(_LIBCPP_HAS_NO_LANGUAGE_SIZED_DEALLOCATION)
-#  define _LIBCPP_HAS_NO_LIBRARY_SIZED_DEALLOCATION
+#if _LIBCPP_STD_VER >= 14 || _LIBCPP_HAS_LANGUAGE_SIZED_DEALLOCATION
```
Comment (ldionne) [inline libcxx/include/new:112]: This is a great example of why this change matters. The "before" code is absolutely impossible to understand, and the "after" code (which is equivalent) is a lot simpler.

File: libcxx/test/support/test_macros.h [hunk 2]
```diff
@@ -214,10 +214,9 @@
 #define TEST_IS_EXECUTED_IN_A_SLOW_ENVIRONMENT
 #endif
 
-#if defined(_LIBCPP_HAS_NO_ALIGNED_ALLOCATION) || \
-  (!(TEST_STD_VER > 14 || \
-    (defined(__cpp_aligned_new) && __cpp_aligned_new >= 201606L)))
-#define TEST_HAS_NO_ALIGNED_ALLOCATION
+#if (!defined(_LIBCPP_HAS_ALIGNED_ALLOCATION) || !_LIBCPP_HAS_ALIGNED_ALLOCATION) ||                                   \
+    (!(TEST_STD_VER > 14 || (defined(__cpp_aligned_new) && __cpp_aligned_new >= 201606L)))
+#  define TEST_HAS_NO_ALIGNED_ALLOCATION
```
Comment (ldionne) [inline libcxx/test/support/test_macros.h:219]: I think I would split this a bit to make it easier to understand. Also, I think it's a bit wrong how it was refactored. Instead:

```
#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_ALIGNED_ALLOCATION
#  define TEST_HAS_NO_ALIGNED_ALLOCATION
#elif TEST_STD_VER < 17 && (!defined(__cpp_aligned_new) || __cpp_aligned_new < 201606L)
#  define TEST_HAS_NO_ALIGNED_ALLOCATION
#endif
```

File: libcxx/test/support/test_macros.h [hunk 3]
```diff
@@ -399,8 +398,8 @@ inline Tp const& DoNotOptimize(Tp const& value) {
 #  define TEST_HAS_OPEN_WITH_WCHAR
 #endif
 
-#if defined(_LIBCPP_HAS_NO_INT128) || defined(_MSVC_STL_VERSION)
-#   define TEST_HAS_NO_INT128
+#if (defined(_LIBCPP_HAS_INT128) && !_LIBCPP_HAS_INT128) || defined(_MSVC_STL_VERSION)
```
Comment (ldionne) [inline libcxx/test/support/test_macros.h:401]:
```suggestion
#if (defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_INT128) || defined(_MSVC_STL_VERSION)
```

File: libcxx/test/support/test_macros.h [hunk 4]
```diff
@@ -419,7 +418,7 @@ inline Tp const& DoNotOptimize(Tp const& value) {
 #  define TEST_HAS_NO_FILESYSTEM
 #endif
 
-#if defined(_LIBCPP_HAS_NO_C8RTOMB_MBRTOC8)
+#if defined(_LIBCPP_HAS_C8RTOMB_MBRTOC8) && !_LIBCPP_HAS_C8RTOMB_MBRTOC8
```
Comment (ldionne) [inline libcxx/test/support/test_macros.h:421]:
```suggestion
#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_C8RTOMB_MBRTOC8
```

**Review by ldionne — COMMENTED** [review]
This basically LGTM with a few comments. Let's try to land this soon to avoid many merge conflicts. I'd still like to have another look once the comments have been addressed.

This simplifies some complicated conditionals *so much*.

File: libcxx/include/__config [hunk 5]
```diff
@@ -318,23 +323,31 @@ typedef __char32_t char32_t;
...
 // Objective-C++ features (opt-in)
 #  if __has_feature(objc_arc)
-#    define _LIBCPP_HAS_OBJC_ARC
+#    define _LIBCPP_HAS_OBJC_ARC 1
+#  else
+#    define _LIBCPP_HAS_OBJC_ARC 0
 #  endif
 
 #  if __has_feature(objc_arc_weak)
 #    define _LIBCPP_HAS_OBJC_ARC_WEAK
```
Comment (ldionne) [inline libcxx/include/__config:332]: Why is this one not a 0-1 macro?

File: libcxx/include/__memory/addressof.h [hunk 6]
```diff
@@ -33,7 +33,7 @@ inline _LIBCPP_HIDE_FROM_ABI __strong _Tp* addressof(__strong _Tp& __x) _NOEXCEP
   return &__x;
 }
 
-#  ifdef _LIBCPP_HAS_OBJC_ARC_WEAK
+#  if _LIBCPP_HAS_OBJC_ARC
```
Comment (ldionne) [inline libcxx/include/__memory/addressof.h:36]: I think this transformation is wrong.

File: libcxx/include/__memory/shared_ptr.h [hunk 7]
```diff
@@ -98,7 +100,7 @@ inline _LIBCPP_HIDE_FROM_ABI _ValueType __libcpp_acquire_load(_ValueType const*
 
 template <class _Tp>
 inline _LIBCPP_HIDE_FROM_ABI _Tp __libcpp_atomic_refcount_increment(_Tp& __t) _NOEXCEPT {
-#if defined(_LIBCPP_HAS_BUILTIN_ATOMIC_SUPPORT) && !defined(_LIBCPP_HAS_NO_THREADS)
+#if _LIBCPP_HAS_BUILTIN_ATOMIC_SUPPORT && !defined(_LIBCPP_HAS_NO_THREADS)
```
Comment (ldionne) [inline libcxx/include/__memory/shared_ptr.h:103]: We need to very quickly switch over to the same scheme for `_LIBCPP_HAS_NO_THREADS` and other things in `__config_site`, because that will become really confusing.

Comment (philnik777) [reply libcxx/include/__memory/shared_ptr.h:103]: I'll work on it as soon as this is landed.
